src/context/TodosProvider.tsx

```diff
@@ -0,0 +1,252 @@
+      } catch (error) {
+        setMessageError(ERROR_MESSAGES[messageId]);
+      }
+    },
+    [setTodos, setMessageError],
+  );
+
+  useEffect(() => {
+    fetchTodos(0);
+  }, [fetchTodos]);
+
+  const handleSubmitSent = useCallback(
+    async (event: React.FormEvent<HTMLFormElement>) => {
+      event.preventDefault();
+
+      // eslint-disable-next-line no-console
+      console.log(query);
+
+      if (!query) {
+        setMessageError(ERROR_MESSAGES[1]);
+        setPending(false);
+
+        return;
+      }
+
+      try {
+        setPending(true);
+        setTempTodo({
+          id: 0,
+          title: query,
+          completed: false,
+          userId: USER_ID,
+        });
+
+        const newTodo = {
+          id: 0,
+          title: query,
+          completed: false,
+          userId: USER_ID,
+        };
+        // eslint-disable-next-line
+        const addedTodo = await addTodo(newTodo);
+
+        setQuery('');
+        setTodos((currentTodo) => [...currentTodo, addedTodo]);
+      } catch (error) {
+        setMessageError(ERROR_MESSAGES[2]);
+      } finally {
+        setMessageError('');
+        setPending(false);
+        setTempTodo(null);
+      }
+    },
+    [setQuery, setMessageError, query],
+  );
+
+  const handleDeleteTodo = useCallback(
+    async (id: number) => {
+      try {
+        setStatus(id);
+        setIsLoadingTodo(todos.find(todo => todo.id === id) || null);
+        setTimeout(() => {
+          setStatus(0);
+          setTodos(
+            currentTodos => currentTodos.filter(todo => todo.id !== id),
+          );
+        }, 500);
+        await deleteTodo(id);
+      } catch (error) {
+        setMessageError(ERROR_MESSAGES[3]);
+      }
+    },
+    [todos],
+  );
+
+  const handleClearCompleted = useCallback(() => {
+    setIsToggled(true);
+
+    setTimeout(async () => {
+      try {
+        const completedTodos = todos.filter((todo) => todo.completed);
+
+        await Promise.all(
+          completedTodos.map(async (todo) => {
+            await handleDeleteTodo(todo.id);
+          }),
+        );
+
+        setTodos(
+          (currentTodos) => currentTodos.filter((todo) => !todo.completed),
+        );
+      } catch (error) {
+        // eslint-disable-next-line no-console
+        console.log(error);
+      } finally {
+        setIsToggled(false);
+      }
+    }, 500);
+  }, [todos, setTodos, setIsToggled, handleDeleteTodo]);
+
+  useMemo(() => {
+    switch (filterType) {
+      case 'All':
+        setFilteredTodos(todos);
+        break;
+      case 'Active':
+        setFilteredTodos(todos.filter(todo => !todo.completed));
+        break;
+      case 'Completed':
+        setFilteredTodos(todos.filter(todo => todo.completed));
+        break;
+      default:
+        break;
+    }
+  }, [filterType, todos, setFilteredTodos]);
+
+  if (messageError) {
+    setTimeout(() => {
+      setMessageError('');
+    }, 3000);
+  }
+
+  const memoizedValue = useMemo(
```

Comment:
This becomes to heavy I think, If everything changes inside memoized value then there is no reason to use `useMemo` at all because in fact we want to retrieve new value on every dependency change